OutOfMemoryError on replication.

Internal Ticket: https://issues.couchbase.com/browse/CBSE-3349

```
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art: Throwing OutOfMemoryError "Failed to allocate a 28 byte allocation with 12 free bytes and 12B until OOM" (recursive case)
01-27 18:14:06.295 7130-7801/com.craig.couchbasesyncexample I/art: WaitForGcToComplete blocked for 123.552ms for cause Alloc
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art: "CBLRequestWorker-http://10.17.1.60:4984/pm-pull-2986f-0" prio=5 tid=23 Runnable
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:   | group="main" sCount=0 dsCount=0 obj=0x22c077c0 self=0xeec3fc00
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:   | sysTid=7807 nice=0 cgrp=default sched=0/0 handle=0xeecb2c00
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:   | state=R schedstat=( 19980811629 1851962167 16071 ) utm=1717 stm=280 core=2 HZ=100
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:   | stack=0xdfb1a000-0xdfb1c000 stackSize=1036KB
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:   | held mutexes= "mutator lock"(shared held)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.lang.String.<init>(String.java:400)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.core.util.TextBuffer.contentsAsString(TextBuffer.java:337)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.core.util.TextBuffer.contentsAsDouble(TextBuffer.java:392)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.core.base.ParserBase._parseSlowFloat(ParserBase.java:818)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.core.base.ParserBase._parseNumericValue(ParserBase.java:797)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.core.base.ParserBase.getDoubleValue(ParserBase.java:714)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:470)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.mapObject(UntypedObjectDeserializer.java:591)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:440)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.mapArray(UntypedObjectDeserializer.java:547)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:454)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:473)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:330)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:25)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3560)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2646)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.support.MultipartDocumentReader.parseJsonBuffer(MultipartDocumentReader.java:63)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.support.MultipartDocumentReader.finish(MultipartDocumentReader.java:122)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.replicator.RemoteBulkDownloaderRequest.finishedPart(RemoteBulkDownloaderRequest.java:228)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.support.MultipartReader.appendData(MultipartReader.java:164)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.replicator.RemoteBulkDownloaderRequest.executeRequest(RemoteBulkDownloaderRequest.java:153)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.replicator.RemoteRequest.execute(RemoteRequest.java:165)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at com.couchbase.lite.replicator.RemoteRequest.run(RemoteRequest.java:105)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
01-27 18:14:06.295 7130-7807/com.craig.couchbasesyncexample E/art:     at java.lang.Thread.run(Thread.java:818)
01-27 18:14:06.372 7130-7141/com.craig.couchbasesyncexample W/art: Suspending all threads took: 77.031ms
01-27 18:14:06.379 7130-7801/com.craig.couchbasesyncexample I/art: Alloc concurrent mark sweep GC freed 943649(27MB) AllocSpace objects, 0(0B) LOS objects, 19% free, 68MB/84MB, paused 809us total 77.403ms
01-27 18:14:06.379 7130-7138/com.craig.couchbasesyncexample I/art: WaitForGcToComplete blocked for 97.049ms for cause Alloc
01-27 18:14:06.390 7130-7141/com.craig.couchbasesyncexample E/System: Uncaught exception thrown by finalizer
01-27 18:14:06.391 7130-7141/com.craig.couchbasesyncexample E/System: java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 12 free bytes and 12B until OOM
                                                                          at com.android.internal.os.BinderInternal$GcWatcher.finalize(BinderInternal.java:51)
                                                                          at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:191)
                                                                          at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:174)
                                                                          at java.lang.Thread.run(Thread.java:818)
01-27 18:14:06.487 7130-7809/com.craig.couchbasesyncexample E/Sync: ExecutionException in Future.get()
                                                                    java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack available
                                                                        at java.util.concurrent.FutureTask.report(FutureTask.java:93)
                                                                        at java.util.concurrent.FutureTask.get(FutureTask.java:163)
                                                                        at com.couchbase.lite.replicator.ReplicationInternal.waitPendingFuturesCompleted(ReplicationInternal.java:1983)
                                                                        at com.couchbase.lite.replicator.PullerInternal.waitForAllTasksCompleted(PullerInternal.java:1075)
                                                                        at com.couchbase.lite.replicator.ReplicationInternal.waitForPendingFutures(ReplicationInternal.java:1929)
                                                                        at com.couchbase.lite.replicator.PullerInternal$14.run(PullerInternal.java:992)
                                                                        at java.lang.Thread.run(Thread.java:818)
                                                                     Caused by: java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack available
```

```
295 7807/ E/art: Throwing OutOfMemoryError "Failed to allocate a 28 byte allocation with 12 free bytes and 12B until OOM" (recursive case)
295 7801/ I/art: WaitForGcToComplete blocked for 123.552ms for cause Alloc
295 7807/ E/art: "CBLRequestWorker-http://10.17.1.60:4984/pm-pull-2986f-0" prio=5 tid=23 Runnable
295 7807/ E/art:   | group="main" sCount=0 dsCount=0 obj=0x22c077c0 self=0xeec3fc00
295 7807/ E/art:   | sysTid=7807 nice=0 cgrp=default sched=0/0 handle=0xeecb2c00
295 7807/ E/art:   | state=R schedstat=( 19980811629 1851962167 16071 ) utm=1717 stm=280 core=2 HZ=100
295 7807/ E/art:   | stack=0xdfb1a000-0xdfb1c000 stackSize=1036KB
295 7807/ E/art:   | held mutexes= "mutator lock"(shared held)
295 7807/ E/art:     at java.lang.String.<init>(String.java:400)
295 7807/ E/art:     at com.fasterxml.jackson.core.util.TextBuffer.contentsAsString(TextBuffer.java:337)
295 7807/ E/art:     at com.fasterxml.jackson.core.util.TextBuffer.contentsAsDouble(TextBuffer.java:392)
295 7807/ E/art:     at com.fasterxml.jackson.core.base.ParserBase._parseSlowFloat(ParserBase.java:818)
295 7807/ E/art:     at com.fasterxml.jackson.core.base.ParserBase._parseNumericValue(ParserBase.java:797)
295 7807/ E/art:     at com.fasterxml.jackson.core.base.ParserBase.getDoubleValue(ParserBase.java:714)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:470)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.mapObject(UntypedObjectDeserializer.java:591)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:440)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.mapArray(UntypedObjectDeserializer.java:547)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:454)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:473)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:330)
295 7807/ E/art:     at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:25)
295 7807/ E/art:     at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3560)
295 7807/ E/art:     at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2646)
295 7807/ E/art:     at com.couchbase.lite.support.MultipartDocumentReader.parseJsonBuffer(MultipartDocumentReader.java:63)
295 7807/ E/art:     at com.couchbase.lite.support.MultipartDocumentReader.finish(MultipartDocumentReader.java:122)
295 7807/ E/art:     at com.couchbase.lite.replicator.RemoteBulkDownloaderRequest.finishedPart(RemoteBulkDownloaderRequest.java:228)
295 7807/ E/art:     at com.couchbase.lite.support.MultipartReader.appendData(MultipartReader.java:164)
295 7807/ E/art:     at com.couchbase.lite.replicator.RemoteBulkDownloaderRequest.executeRequest(RemoteBulkDownloaderRequest.java:153)
295 7807/ E/art:     at com.couchbase.lite.replicator.RemoteRequest.execute(RemoteRequest.java:165)
295 7807/ E/art:     at com.couchbase.lite.replicator.RemoteRequest.run(RemoteRequest.java:105)
295 7807/ E/art:     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
295 7807/ E/art:     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
295 7807/ E/art:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
295 7807/ E/art:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
295 7807/ E/art:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
295 7807/ E/art:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
295 7807/ E/art:     at java.lang.Thread.run(Thread.java:818)
372 7141/ W/art: Suspending all threads took: 77.031ms
379 7801/ I/art: Alloc concurrent mark sweep GC freed 943649(27MB) AllocSpace objects, 0(0B) LOS objects, 19% free, 68MB/84MB, paused 809us total 77.403ms
379 7138/ I/art: WaitForGcToComplete blocked for 97.049ms for cause Alloc
390 7141/ E/System: Uncaught exception thrown by finalizer
391 7141/ E/System: java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 12 free bytes and 12B until OOM
                    at com.android.internal.os.BinderInternal$GcWatcher.finalize(BinderInternal.java:51)
                    at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:191)
                    at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:174)
                    at java.lang.Thread.run(Thread.java:818)
487 7809/ E/Sync: ExecutionException in Future.get()
              java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack available
                  at java.util.concurrent.FutureTask.report(FutureTask.java:93)
                  at java.util.concurrent.FutureTask.get(FutureTask.java:163)
                  at com.couchbase.lite.replicator.ReplicationInternal.waitPendingFuturesCompleted(ReplicationInternal.java:1983)
                  at com.couchbase.lite.replicator.PullerInternal.waitForAllTasksCompleted(PullerInternal.java:1075)
                  at com.couchbase.lite.replicator.ReplicationInternal.waitForPendingFutures(ReplicationInternal.java:1929)
                  at com.couchbase.lite.replicator.PullerInternal$14.run(PullerInternal.java:992)
                  at java.lang.Thread.run(Thread.java:818)
               Caused by: java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack available
```

--- a/src/main/java/com/couchbase/lite/internal/Body.java
+++ b/src/main/java/com/couchbase/lite/internal/Body.java
@@ -31,6 +31,7 @@ import java.util.Map;
  */
 public class Body {
     private byte[] json;
+    private long size = 0;
     private Object object;
 
     public Body(byte[] json) {
@@ -39,6 +40,16 @@ public class Body {
 
     public Body(Map<String, Object> properties) {
         this.object = properties;
+    }
+
+    public Body(byte[] json, long size) {
+        this.json = json;
+        this.size = size;
+    }
+
+    public Body(Map<String, Object> properties, long size) {
+        this.object = properties;
+        this.size = size;
     }
 
     public Body(List<?> array) {
@@ -170,5 +181,9 @@ public class Body {
 
     public Object getObject(String key) {
         return getProperties() != null ? getProperties().get(key) : null;
+    }
+
+    public long getSize() {
+        return size;
     }
 }

--- a/src/main/java/com/couchbase/lite/internal/RevisionInternal.java
+++ b/src/main/java/com/couchbase/lite/internal/RevisionInternal.java
@@ -64,6 +64,10 @@ public class RevisionInternal {
         this(new Body(properties));
     }
 
+    public RevisionInternal(Map<String, Object> properties, long size) {
+        this(new Body(properties, size));
+    }
+
     ////////////////////////////////////////////////////////////
     // Setter / Getter methods
     ////////////////////////////////////////////////////////////

--- a/src/main/java/com/couchbase/lite/replicator/PulledRevision.java
+++ b/src/main/java/com/couchbase/lite/replicator/PulledRevision.java
@@ -40,6 +40,10 @@ class PulledRevision extends RevisionInternal {
         super(properties);
     }
 
+    public PulledRevision(Map<String, Object> properties, long size) {
+        super(properties, size);
+    }
+
     public String getRemoteSequenceID() {
         return remoteSequenceID;
     }

--- a/src/main/java/com/couchbase/lite/replicator/PullerInternal.java
+++ b/src/main/java/com/couchbase/lite/replicator/PullerInternal.java
@@ -47,6 +47,7 @@ import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
 
 import okhttp3.OkHttpClient;
 import okhttp3.Response;
@@ -77,6 +78,8 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
     private static final int INSERTION_BATCHER_DELAY = 250; // 0.25 Seconds
     private static final int INSERTION_BATCHER_CAPACITY = 100;
 
+    private static final long MAX_QUEUE_MEMORY_SIZE = 2 * 1024 * 1024; // 2MB
+
     private ChangeTracker changeTracker;
     protected SequenceMap pendingSequences;
     protected Boolean canBulkGet;  // Does the server support _bulk_get requests?
@@ -88,6 +91,7 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
             new ArrayList<RevisionInternal>(100));
     protected int httpConnectionCount;
     protected Batcher<RevisionInternal> downloadsToInsert;
+    protected AtomicLong queuedMemorySize = new AtomicLong(0);
 
     private String str = null;
 
@@ -125,6 +129,9 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
                 @Override
                 public void process(List<RevisionInternal> inbox) {
                     insertDownloads(inbox);
+                    if (downloadsToInsert.count() == 0) {
+                        queuedMemorySize.set(0);
+                    }
                 }
             });
         }
@@ -332,12 +339,12 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
                     db,
                     this.requestHeaders,
                     new RemoteBulkDownloaderRequest.BulkDownloaderDocument() {
-                        public void onDocument(Map<String, Object> props) {
+                        public void onDocument(Map<String, Object> props, long size) {
                             // Got a revision!
                             // Find the matching revision in 'remainingRevs' and get its sequence:
                             RevisionInternal rev;
                             if (props.get("_id") != null) {
-                                rev = new RevisionInternal(props);
+                                rev = new RevisionInternal(props, size);
                             } else {
                                 rev = new RevisionInternal((String) props.get("id"),
                                         (String) props.get("rev"), false);
@@ -456,12 +463,22 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
             }
         }
 
-        if (rev != null && rev.getBody() != null)
-            rev.getBody().compact();
+        // NOTE: should not/not necessary to call Body.compact()
+        // new RevisionInternal(Map<string, Object>) creates Body instance only
+        // with `object`. Serializing object to json causes two unnecessary
+        // JSON serializations.
+
+        if (rev.getBody() != null)
+            queuedMemorySize.addAndGet(rev.getBody().getSize());
 
         downloadsToInsert.queueObject(rev);
-    }
 
+        // if queue memory size is more than maximum, force flush the queue.
+        if (queuedMemorySize.get() > MAX_QUEUE_MEMORY_SIZE) {
+            Log.d(TAG, "Flushing queued memory size at: " + queuedMemorySize);
+            downloadsToInsert.flushAll(true);
+        }
+    }
 
     // Get as many revisions as possible in one _all_docs request.
     // This is compatible with CouchDB, but it only works for revs of generation 1 without attachments.
@@ -586,7 +603,9 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
                                 continue;
                             }
                         }
-                        if (rev.getBody() != null) rev.getBody().compact();
+
+                        // NOTE: calling Body.compact() here cause another JSON serialization by Jackson.
+                        //       At this point Body.json is null, and Body.object has values.
 
                         // Mark this revision's fake sequence as processed:
                         pendingSequences.removeSequence(fakeSequence);
@@ -700,18 +719,33 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
                                 setError(e);
                             }
                         } else {
+
                             Map<String, Object> properties = (Map<String, Object>) result;
-                            PulledRevision gotRev = new PulledRevision(properties);
+                            long size = 0;
+                            if (httpResponse != null && httpResponse.body() != null)
+                                size = httpResponse.body().contentLength();
+                            PulledRevision gotRev = new PulledRevision(properties, size);
                             gotRev.setSequence(rev.getSequence());
 
                             Log.d(TAG, "%s: pullRemoteRevision add rev: %s to batcher: %s",
                                     PullerInternal.this, gotRev, downloadsToInsert);
 
+                            // NOTE: should not/not necessary to call Body.compact()
+                            // new PulledRevision(Map<string, Object>) creates Body instance only
+                            // with `object`. Serializing object to json causes two unnecessary
+                            // JSON serializations.
+
                             if (gotRev.getBody() != null)
-                                gotRev.getBody().compact();
+                                queuedMemorySize.addAndGet(gotRev.getBody().getSize());
 
                             // Add to batcher ... eventually it will be fed to -insertRevisions:.
                             downloadsToInsert.queueObject(gotRev);
+
+                            // if queue memory size is more than maximum, force flush the queue.
+                            if (queuedMemorySize.get() > MAX_QUEUE_MEMORY_SIZE) {
+                                Log.d(TAG, "Flushing  queued memory size at: " + queuedMemorySize);
+                                downloadsToInsert.flushAll(true);
+                            }
                         }
 
                         // Note that we've finished this task:

--- a/src/main/java/com/couchbase/lite/replicator/RemoteBulkDownloaderRequest.java
+++ b/src/main/java/com/couchbase/lite/replicator/RemoteBulkDownloaderRequest.java
@@ -49,7 +49,7 @@ public class RemoteBulkDownloaderRequest extends RemoteRequest implements Multip
 
     @InterfaceAudience.Private
     public interface BulkDownloaderDocument {
-        void onDocument(Map<String, Object> props);
+        void onDocument(Map<String, Object> props, long size);
     }
 
     ////////////////////////////////////////////////////////////
@@ -226,7 +226,7 @@ public class RemoteBulkDownloaderRequest extends RemoteRequest implements Multip
         if (_docReader == null)
             throw new IllegalStateException("_docReader is not defined");
         _docReader.finish();
-        _onDocument.onDocument(_docReader.getDocumentProperties());
+        _onDocument.onDocument(_docReader.getDocumentProperties(), _docReader.getDocumentSize());
         _docReader = null;
         Log.v(TAG, "%s: Finished document", this);
     }

--- a/src/main/java/com/couchbase/lite/support/MultipartDocumentReader.java
+++ b/src/main/java/com/couchbase/lite/support/MultipartDocumentReader.java
@@ -33,6 +33,7 @@ public class MultipartDocumentReader implements MultipartReaderDelegate {
     private CustomByteArrayOutputStream jsonBuffer;
     private boolean jsonCompressed;
     private Map<String, Object> document;
+    private long documentSize = 0; // original JSON document size
     private Database database;
     private Map<String, BlobStoreWriter> attachmentsByName;
     private Map<String, BlobStoreWriter> attachmentsByMd5Digest;
@@ -45,7 +46,14 @@ public class MultipartDocumentReader implements MultipartReaderDelegate {
         return document;
     }
 
+    public long getDocumentSize() {
+        return documentSize;
+    }
+
     public void parseJsonBuffer() {
+
+        documentSize = jsonBuffer.count();
+
         ByteArrayInputStream in = new ByteArrayInputStream(jsonBuffer.buf(), 0, jsonBuffer.count());
         try {
             try {

